src/qibolab/sequence.py

```diff
@@ -93,17 +94,69 @@ def pulse_channels(self, pulse_id: int) -> list[ChannelId]:
         """Find channels on which a pulse with a given id plays."""
         return [channel for channel, pulse in self if pulse.id == pulse_id]
 
-    def concatenate(self, other: "PulseSequence") -> None:
+    def concatenate(self, other: Iterable[_Element]) -> None:
+        """Concatenate two sequences.
+
+        Appends ``other`` in-place such that the result is:
+            - ``self``
+            - necessary delays to synchronize channels
+            - ``other``
+        Guarantees that the all the channels in the concatenated
+        sequence will start simultaneously
+        """
+        _synchronize(self, PulseSequence(other).channels)
+        self.extend(other)
+
+    def __ilshift__(self, other: Iterable[_Element]) -> "PulseSequence":
+        """Juxtapose two sequences.
+
+        Alias to :meth:`concatenate`.
+        """
+        self.concatenate(other)
+        return self
+
+    def __lshift__(self, other: Iterable[_Element]) -> "PulseSequence":
+        """Juxtapose two sequences.
+
+        A copy is made, and no input is altered.
+
+        Other than that, it is based on :meth:`concatenate`.
+        """
+        copy = deepcopy(self)
+        copy <<= other
+        return copy
+
+    def juxtapose(self, other: Iterable[_Element]) -> None:
         """Juxtapose two sequences.
 
         Appends ``other`` in-place such that the result is:
             - ``self``
             - necessary delays to synchronize channels
             - ``other``
+        Guarantee simultaneous start and no overlap.
         """
-        _synchronize(self, other.channels)
+        _synchronize(self, PulseSequence(other).channels | self.channels)
         self.extend(other)
 
+    def __ior__(self, other: Iterable[_Element]) -> "PulseSequence":
+        """Juxtapose two sequences.
+
+        Alias to :meth:`concatenate`.
+        """
+        self.juxtapose(other)
+        return self
+
+    def __or__(self, other: Iterable[_Element]) -> "PulseSequence":
+        """Juxtapose two sequences.
+
+        A copy is made, and no input is altered.
+
+        Other than that, it is based on :meth:`concatenate`.
+        """
+        copy = deepcopy(self)
```

Comment:
On a second thought, @alecandido what is the main motivation of using `deepcopy` here? For sure we need a `copy` of the sequence, because we don't want to change the original one, however a `deepcopy` will also create copies of all the pulses in the original sequence, for example in
```py
s = s1 | s2
```
`s` will contain copies (with new ids) of all pulses from `s1` but the same pulses (same ids) from `s2`, which is a bit assymetric.

I believe it would be a bit more flexible with `copy` (reuse the original pulses of `s1`), because if the user wants to specifically create new pulses, they can always `deepcopy` the result on their own. While if we `deepcopy` internally, we don't give them the option to use the same pulses at all.

But maybe I am missing something.

Reply:
I believe you're getting the logic: IDs are associated to objects. If the copy is not deep, then you will have the same IDs.

The point is that this is not something that is apparent to the user, because you're just appending gates.
Let's take one particular example:
```py
# native gates for qubit 0
q0 = ...
seq = q0.RX() | q0.M() | q0.RX() | q0.M()
```
this is a perfectly valid sequence (possibly you're manually unrolling...), but it can not be executed, because of the conflicting IDs.
Then, what you should do is:
```py
seq = q0.RX() | deepcopy(q0.M()) | q0.RX() | deepcopy(q0.M())
```
which is much less intuitive.

So, since this notation is just meant as a simplified (somewhat higher-level) interface, I decided to enforce in the library the uniqueness of the IDs.
I acknowledge it's debatable.

Reply:
`deepcopy()` removed in 2d6a018